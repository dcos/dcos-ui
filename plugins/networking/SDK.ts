let SDK;

module.exports = {
  getSDK() {
    return SDK;
  },
  setSDK(pluginSDK) {
    SDK = pluginSDK;
  }
};
