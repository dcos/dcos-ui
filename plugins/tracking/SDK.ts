let SDK;

export default {
  getSDK() {
    return SDK;
  },
  setSDK(pluginSDK) {
    SDK = pluginSDK;
  }
};
