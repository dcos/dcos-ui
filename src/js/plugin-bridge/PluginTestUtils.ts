import PluginSDK from "PluginSDK";

/**
 * Finds and returns the SDK for a specific plugin
 * @param  {String} pluginID - ID of plugin
 * @param  {Object} config   - configuration
 * @return {PluginSDK}          - SDK for plugin with pluginID
 */
function getSDK(pluginID, config) {
  // Get SDK for pluginID.
  return PluginSDK.__getSDK(pluginID, config);
}

/**
 * Add reducer to Store. Could be an actual plugin reducer or a mock for test
 * cases that require specific state.
 * @param {String} root    - Root key for which reducer will manage state
 * @param {Function} reducer - Reducer function
 */
function addReducer(root, reducer) {
  PluginSDK.__addReducer(root, reducer);
}

const TestUtils = {
  addReducer,
  getSDK
};

export default TestUtils;
