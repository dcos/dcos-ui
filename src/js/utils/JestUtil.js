import PropTypes from "prop-types";
import React from "react";
import { routerShape } from "react-router";
import { I18nProvider } from "@lingui/react";

import en from "#LOCALE/en/messages.js";

// Private router stub
const RouterStub = {
  push() {},
  replace() {},
  go() {},
  goBack() {},
  goForward() {},
  setRouteLeaveHook() {},
  createPath() {},
  createHref() {},
  isActive() {}
};

// Default prototype functions when mocking timezone
const defaultGetTimezoneOffset = Date.prototype.getTimezoneOffset;
const defaultToLocaleString = Date.prototype.toLocaleString;
const defaultDateTimeFormat = Intl.DateTimeFormat;

const JestUtil = {
  /**
   * Generates a callback function to a filter() call that will
   * keep only DOMElements matching the given tag name(s)
   *
   * @example <caption>How to use filterByTagName</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var nameTDs = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance, 'table-column-name'
   * ).filter(filterByTagName('TD'));
   *
   * @param {array|string} tag - One or mor tagNames to match
   * @returns {function} Returns a callback function to be passed on .filter()
   */
  filterByTagName(tag) {
    if (!Array.isArray(tag)) {
      tag = [tag];
    }

    return function(element) {
      return tag.some(function(tagName) {
        return element.tagName === tagName;
      });
    };
  },

  /**
   * Helper callback function to be used when you want the text contents
   * of an array of DOMElements.
   *
   * @example <caption>How to use mapTextContent</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var nameTDContents = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance,
   *   'table-column-name'
   * ).map(mapTextContent);
   *
   * @param {DOMElement} element - The DOM element for which to get the content
   * @returns {string} The text content
   */
  mapTextContent(element) {
    return element.textContent.trim();
  },

  /**
   * Mock a different timezone by overriding relevant Date primitive
   * prototype functions.
   *
   * This function will mock Date.getTimezoneOffset, Date.toLocaleString and Intl.DateTimeFormat
   *
   * @param {String} timezone - The IANA timezone string (ex. Europe/Athens) or 'UTC'
   */
  mockTimezone(timezone) {
    const date = new Date();
    const timezoneOffset =
      (new Date(date.toLocaleString(undefined, { timeZone: "UTC" })) -
        new Date(date.toLocaleString(undefined, { timeZone: timezone }))) /
      1000 /
      60;

    /* eslint-disable no-extend-native */
    Date.prototype.getTimezoneOffset = function() {
      return timezoneOffset;
    };
    Date.prototype.toLocaleString = function(locale = undefined, options = {}) {
      options.timeZone = options.timeZone || timezone;

      return defaultToLocaleString.call(this, locale, options);
    };
    Intl.DateTimeFormat = function(locales, options) {
      options.timeZone = options.timeZone || timezone;

      return defaultDateTimeFormat.call(this, locales, options);
    };
    /* eslint-enable no-extend-native */
  },

  /**
   * Helper to stub router context, based on
   * https://github.com/reactjs/react-router/blob/0.13.x/docs/guides/testing.md
   * @param {React.Component} Component
   * @param {object} [routerStubs]
   * @param {object} [contextTypes]
   * @param {object} [context]
   * @returns {React.Element} wrapped component element
   */
  stubRouterContext(
    Component,
    routerStubs = {},
    contextTypes = {},
    context = {}
  ) {
    // Create wrapper component
    return class WrappedComponent extends React.Component {
      static get childContextTypes() {
        return Object.assign(
          {
            router: routerShape,
            routeDepth: PropTypes.number
          },
          contextTypes
        );
      }

      getChildContext() {
        return Object.assign(
          {
            router: Object.assign(RouterStub, routerStubs),
            routeDepth: 0
          },
          context
        );
      }

      render() {
        return <Component {...this.props} />;
      }
    };
  },

  withI18nProvider(Component, catalogs = { en }) {
    // eslint-disable-next-line react/no-multi-comp
    return class WrappedComponent extends React.Component {
      render() {
        return (
          <I18nProvider defaultRender="span" language="en" catalogs={catalogs}>
            <Component {...this.props} />
          </I18nProvider>
        );
      }
    };
  },

  /**
   * Generates a callback function to a reduce() call that will
   * find and return the textContent of DOM element(s) matching
   * the given selector
   *
   * @example <caption>How to use reduceTextContentOfSelector</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var names = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance,
   *   'table-column-name'
   * ).reduce(
   *   reduceTextContentOfSelector('.collapsing-string-full-string'),
   *   []
   * );
   *
   * @param {string} selector - The CSS selector to match children
   * @returns {function} Returns a callback function to be passed on .reduce()
   */
  reduceTextContentOfSelector(selector) {
    return function(strings, element) {
      const matchedElements = Array.from(element.querySelectorAll(selector));

      matchedElements.forEach(function(stringElement) {
        strings.push(stringElement.textContent);
      });

      return strings;
    };
  },

  /**
   * Restore the original version of the Date prototype functions, overwritten
   * by the mockTimezone function.
   */
  unmockTimezone() {
    /* eslint-disable no-extend-native */
    Date.prototype.getTimezoneOffset = defaultGetTimezoneOffset;
    Date.prototype.toLocaleString = defaultToLocaleString;
    Intl.DateTimeFormat = defaultDateTimeFormat;
    /* eslint-enable no-extend-native */
  }
};

module.exports = JestUtil;
