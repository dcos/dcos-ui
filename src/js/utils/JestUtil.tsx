import PropTypes from "prop-types";
import * as React from "react";
import { routerShape } from "react-router";
import { I18nProvider } from "@lingui/react";

import en from "#LOCALE/en/messages.js";

// Private router stub
const RouterStub = {
  push() {},
  replace() {},
  go() {},
  goBack() {},
  goForward() {},
  setRouteLeaveHook() {},
  createPath() {},
  createHref() {},
  isActive() {},
};

// Default prototype functions when mocking timezone
const defaultGetTimezoneOffset = Date.prototype.getTimezoneOffset;
const defaultToLocaleString = Date.prototype.toLocaleString;
const defaultDateTimeFormat = Intl.DateTimeFormat;

const JestUtil = {
  /**
   * Mock a different timezone by overriding relevant Date primitive
   * prototype functions.
   *
   * This function will mock Date.getTimezoneOffset, Date.toLocaleString and Intl.DateTimeFormat
   *
   * @param {String} timezone - The IANA timezone string (ex. Europe/Athens) or 'UTC'
   */
  mockTimezone(timezone) {
    const date = new Date();
    const timezoneOffset =
      (new Date(date.toLocaleString(undefined, { timeZone: "UTC" })) -
        new Date(date.toLocaleString(undefined, { timeZone: timezone }))) /
      1000 /
      60;

    Date.prototype.getTimezoneOffset = () => timezoneOffset;
    Date.prototype.toLocaleString = function (
      locale = undefined,
      options = {}
    ) {
      options.timeZone = options.timeZone || timezone;

      return defaultToLocaleString.call(this, locale, options);
    };
    Intl.DateTimeFormat = function (locales, options) {
      options.timeZone = options.timeZone || timezone;

      return defaultDateTimeFormat.call(this, locales, options);
    };
  },

  /**
   * Helper to stub router context, based on
   * https://github.com/reactjs/react-router/blob/0.13.x/docs/guides/testing.md
   * @param {React.Component} Component
   * @param {object} [routerStubs]
   * @param {object} [contextTypes]
   * @param {object} [context]
   * @returns {React.Element} wrapped component element
   */
  stubRouterContext(
    Component,
    routerStubs = {},
    contextTypes = {},
    context = {}
  ) {
    // Create wrapper component
    return class WrappedComponent extends React.Component {
      static get childContextTypes() {
        return {
          router: routerShape,
          routeDepth: PropTypes.number,
          ...contextTypes,
        };
      }

      getChildContext() {
        return {
          router: { ...RouterStub, ...routerStubs },
          routeDepth: 0,
          ...context,
        };
      }

      render() {
        return <Component {...this.props} />;
      }
    };
  },

  withI18nProvider(Component, catalogs = { en }) {
    return class WrappedComponent extends React.Component {
      render() {
        return (
          <I18nProvider defaultRender="span" language="en" catalogs={catalogs}>
            <Component {...this.props} />
          </I18nProvider>
        );
      }
    };
  },

  /**
   * Restore the original version of the Date prototype functions, overwritten
   * by the mockTimezone function.
   */
  unmockTimezone() {
    Date.prototype.getTimezoneOffset = defaultGetTimezoneOffset;
    Date.prototype.toLocaleString = defaultToLocaleString;
    Intl.DateTimeFormat = defaultDateTimeFormat;
  },
};

export default JestUtil;
