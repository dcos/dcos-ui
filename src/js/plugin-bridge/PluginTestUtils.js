import JestUtil from '../utils/JestUtil';
import Loader from './Loader';
import PluginModules from './PluginModules';
import PluginSDK from 'PluginSDK';

/**
 * Loads whatever plugins are passed in. Could be Mocks
 * @param  {Object} plugins - Plugin name to Config
 */
function loadPlugins(plugins) {
  var availablePlugins = {};
  var pluginConfig = {};

  Object.keys(plugins).forEach(function (pluginID) {
    availablePlugins[pluginID] = plugins[pluginID].module;
    pluginConfig[pluginID] = plugins[pluginID].config;
  });

  Loader.__setMockPlugins(availablePlugins);
  PluginSDK.initialize(pluginConfig);
}

/**
 * Finds actual plugins by name and loads them
 * @param  {Object} plugins - Map of name to config
 */
function loadPluginsByName(plugins) {
  let pluginsToLoad = {};

  let {pluginsList, externalPluginsList} = Loader.__getAvailablePlugins();

  Object.keys(plugins).forEach(function (pluginID) {

    if (!(pluginID in pluginsList) && !(pluginID in externalPluginsList)) {
      throw new Error(`${pluginID} does not exist. Failed to load.`);
    }
    let path;
    // Default to always selecting internal plugin if same pluginID
    // exists in external plugins. This makes mocking easier.
    if (pluginID in pluginsList) {
      path = pluginsList[pluginID];
    } else {
      path = externalPluginsList[pluginID];
    }
    pluginsToLoad[pluginID] = {
      module: path,
      config: plugins[pluginID]
    };
  });
  loadPlugins(pluginsToLoad);

}

/**
 * Finds and returns the SDK for a specific plugin
 * @param  {String} pluginID - ID of plugin
 * @param  {Object} config   - configuration
 * @param  {Boolean} loadPlugin
 * @return {PluginSDK}          - SDK for plugin with pluginID
 */
function getSDK(pluginID, config, loadPlugin = false) {
  // Load plugin first so it can register it's reducer
  // if it has one.
  if (loadPlugin) {
    loadPluginsByName({
      [pluginID]: config
    });
  }
  // Get SDK for pluginID.
  return PluginSDK.__getSDK(pluginID, config);
}

/**
 * Set a mock for a module which would normally be returned by
 * PluginSDK.get().
 * @param {String} name - name of module
 * @param {any} mock - value representing the mock
 * @returns {mock} - the mock that was passed in
 */
function setMock(name, mock) {
  Loader.__setMockModule(name, mock);
  return mock;
}

/**
 * Add reducer to Store. Could be an actual plugin reducer or a mock for test
 * cases that require specific state.
 * @param {String} root    - Root key for which reducer will manage state
 * @param {Function} reducer - Reducer function
 */
function addReducer(root, reducer) {
  PluginSDK.__addReducer(root, reducer);
}

/**
 * Takes an Array or String representing a module name{s) that need unMocking,
 * finds the directory of the file and calls jest.dontMock on path.
 * @param  {Array|String} moduleNames - Names of modules to dontMock

 */
function dontMock(moduleNames) {
  if (Array.isArray(moduleNames)) {
    moduleNames.forEach(dontMock);
    return;
  }
  // Just one module to mock
  var name = moduleNames;
  // Try unmocking store first
  if (JestUtil.dontMockStore(name)) {
    return;
  }
  // Assuming modules have unique names
  let foundType = Object.keys(PluginModules).filter(moduleType => {
    return name in PluginModules[moduleType];
  });
  if (!foundType.length) {
    throw new Error(`Module ${name} does not exist.`);
  }
  let modulePath = `../${foundType[0]}/${PluginModules[foundType[0]][name]}`;
  jest.dontMock(modulePath);
}

let TestUtils = {
  addReducer,
  getSDK,
  dontMock,
  loadPlugins,
  loadPluginsByName,
  setMock
};

module.exports = TestUtils;
