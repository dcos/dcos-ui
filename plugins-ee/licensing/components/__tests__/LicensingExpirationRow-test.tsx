import * as React from "react";
import { mount } from "enzyme";
import JestUtil from "#SRC/js/utils/JestUtil";

import PluginTestUtils from "PluginTestUtils";

const SDK = PluginTestUtils.getSDK("licensing", { enabled: true });
require("../../SDK").setSDK(SDK);

const LicensingStore = require("../../stores/LicensingStore").default;
const LicensingExpirationRow = require("../LicensingExpirationRow").default;
const LicensingSummary = require("../../structs/LicensingSummary").default;

const renderer = require("react-test-renderer");

const licensingFixture = require("../../../../tests/_fixtures/licensing/licensing-summary");

const WrappedComponent = JestUtil.withI18nProvider(LicensingExpirationRow);

describe("LicensingExpirationRow", () => {
  describe("#render", () => {
    it("renders nothing when no data available", () => {
      spyOn(LicensingStore, "getLicensingSummary").and.returnValue(null);
      const expirationRow = mount(<WrappedComponent />);

      expect(expirationRow.find(".configuration-map-row").length).toBe(0);
    });

    it("renders days remaining message when not expired", () => {
      const summary = new LicensingSummary(licensingFixture);
      spyOn(LicensingStore, "getLicensingSummary").and.returnValue(summary);
      const instance = mount(<WrappedComponent />);

      const toggleValue = instance.find(".configuration-map-value");
      expect(toggleValue.text()).toContain("days left");
    });

    it("renders expired message when expired", () => {
      const copy = {
        ...licensingFixture
      };
      copy.current_timestamp = "2020-01-26T10:05:00+04:00";
      const summary = new LicensingSummary(copy);
      spyOn(LicensingStore, "getLicensingSummary").and.returnValue(summary);
      const instance = mount(<WrappedComponent />);

      expect(instance.find(".configuration-map-value").text()).toContain(
        "days overdue"
      );
    });

    it("renders entire expiration row", () => {
      const summary = new LicensingSummary(licensingFixture);
      spyOn(LicensingStore, "getLicensingSummary").and.returnValue(summary);

      const instance = renderer.create(<WrappedComponent />);

      const tree = instance.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
