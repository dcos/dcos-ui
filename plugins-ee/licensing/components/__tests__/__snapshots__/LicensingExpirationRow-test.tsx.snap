// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicensingExpirationRow #render renders entire expiration row 1`] = `
<div
  className="configuration-map-row table-row"
>
  <div
    className="configuration-map-label"
  />
  <div
    className="configuration-map-value"
  >
    <span
      className="toggle-value"
      onClick={[Function]}
    >
      862 days left
    </span>
  </div>
  <a
    className="button button-primary-link text-no-transform"
    href="https://docs.mesosphere.com/latest/administering-clusters/licenses"
    target="_blank"
  />
</div>
`;
